apache/cloudstack
# Auto Enable/Disable KVM hosts

### Description

This PR allows CloudStack to auto-disable and auto-enable KVM hosts resource state based on customisable host/hypervisor health checks.

#### Management server changes

A new cluster configuration 'enable.kvm.host.auto.enable.disable' is added, disabled by default.

This PR extends the AgentManager to process StartupRoutingCommand and PingRoutingCommand with the host health check result:
 - If the host health check result is null, do nothing.
 - If the host health check result is true, enable the host resource state if it is Disabled.
 - If the host health check result is false, disable the host resource state if it is Enabled.

On every automatic enable or disable event, the management server will send an alert to the admin and add/update an automatic annotation (comment) on the specific host.

- If a host gets auto-disabled by a health check failure, then it can be auto-enabled when the health check succeeds. But if the host gets disabled by the admin, then it must not be auto-enabled when the health check succeeds (manual host disabling takes precedence over the auto-enabling of a host).
- If a host gets auto-disabled by a health check failure, the admin could enable the host but unless they also disable the health check on the host then it will just get disabled again when the health check fails

CloudStack controls when a host can/cannot be auto-enabled or auto-disabled by a host detail record (on the host_details table) with key ‘autoenablekvmhost’, in the following way:
- If a host is auto-enabled or auto-disabled and there is no host detail with key ‘autoenablekvmhost’ for that host, then a new host detail record is created with the key ‘autoenablekvmhost’ and is set to ‘true’. (indicating that the host can be auto-enabled/auto-disabled)
- If the administrator manually disables a host and there is a host detail with key ‘autoenablekvmhost’ for that host, then the host detail ‘autoenablekvmhost’ is set to ‘false’ (indicating that the host cannot be auto-enabled if the health check succeeds)
- If the administrator manually enables a host and there is a host detail with key ‘autoenablekvmhost’ for that host, then the host detail ‘autoenablekvmhost’ is set to ‘true’ (indicating that the host can be auto-disabled if the health check fails)
- If the cluster setting is disabled, and the administrator enables/disables hosts in the cluster, then the host detail with key ‘autoenablekvmhost’ is not created for the hosts. (preserving the usual behavior)


#### KVM host Health Check

Set the path of an executable file/script for host health check on the property 'agent.health.check.script.path' on `agent.properties` 

- The health check result is true, if the script is executed successfully and the exit code is 0
- The health check result is false, if the script is executed successfully and the exit code is 1
- The health check result is null, if 
   - Script file is not specified, or
   - Script file does not exist, or 
   - Script file is not accessible by the user of the cloudstack-agent process, or
   - Script file is not executable
   - There are errors when the script is executed (exit codes other than 0 or 1)

### Types of changes


### Feature/Enhancement Scale or Bug Severity

#### Feature/Enhancement Scale


#### Bug Severity



### Screenshots (if appropriate):


### How Has This Been Tested?
2xRocky8 KVM hosts in 1 cluster -> put host1 into maintenance, test host2
- Enable the cluster setting wihout providing a health script to the host -> no automatic enable/disable
- Enable the cluster setting, set a script that doesn't have permission rights -> no automatic enable/disable
- Enable the cluster setting, set a script that succeeds (ping the management server) -> observe the health check result is successful, but since the host is Enabled is not transitioned to any state
- Enable the cluster setting, set a script that fails (ping an unreacheable host) -> observe the health check result is false, the host2 is put into Disabled state
- Enable the host2 manually -> the host2 is set to Enabled state
- The next failure sets the hosts2 into Disabled state again
- Modify the script again so it succeeds, the host2 goes into Enabled state, then disable manually the host2 -> the future successful health checks do not put the host into Disabled state
